fpt/src/memory.rs

```diff
@@ -260,11 +271,17 @@ impl Bus {
     pub fn load_bootrom(&mut self) {
         let bootrom = self.memory().bootrom;
         self.clone_from_slice(map::BOOTROM, bootrom);
+        for i in map::BOOTROM {
+            self.memory_mut().code_listing[i] = None;
+        }
     }
 
     pub fn unload_bootrom(&mut self) {
-        let cartridge = self.memory_mut().cartridge[0x0000..0x0100].to_vec();
+        let cartridge = self.memory_mut().cartridge[map::BOOTROM].to_vec();
         self.clone_from_slice(map::BOOTROM, &cartridge);
```

Comment:
Slightly unrelated:

The borrow checker isn't allowing us to clone a slice from the cartridge array directly to the memory array, right?

I think I'm finally realizing how we could solve this sort of lifetime issues: by moving most of `Bus`'s auxiliary functions up to `Memory` (so that they wouldn't have to "lock" the RefCell) and providing just a single function:

```rust
pub fn borrow<R>(borrower: impl FnOnce(mem: &Memory) -> R) -> R {
    borrower(&self.0.borrow())
}

pub fn borrow_mut…
```

Maybe we could even rename `struct Memory` to `struct BusImpl` and call this a pImpl, idk

But I don't know if this makes much sense in our case, this is copying what egui does (but this seems to be a pattern in Rust).

No, it doesn't. In egui, the client only needs to use pieces of the Context in an isolated manner. This doesn't solve the case where we want to lock on multiple pieces of hidden state.

Wait, we could also just use our accessors `memory` and `memory_mut` if the helpers from `Bus` were moved to `Memory`.

Is this supposed to be hard?? I wonder if one `unsafe` or two could drastically simplify this API...

Reply:
bonk no unsafe!!!

Reply:
You're right, yesterday I learned about _reborrowing_. Now I know how to share sub-parts of an object from a `RefCell`'s `RefMut`.